io-uring-test/src/tests/net.rs

```diff
@@ -473,34 +473,158 @@ pub fn test_tcp_buffer_select<S: squeue::EntryMarker, C: cqueue::EntryMarker>(
         _ => panic!("{}", cqe.flags()),
     }
 
+    // provide buf for recvmsg
+    let mut recvmsg_buf = [0u8; 1024];
+    let provide_bufs_e = opcode::ProvideBuffers::new(recvmsg_buf.as_mut_ptr(), 1024, 1, 0xdeaf, 0);
+
+    unsafe {
+        ring.submission()
+            .push(&provide_bufs_e.build().user_data(0x26).into())
+            .expect("queue is full");
+    }
+
+    ring.submit_and_wait(1)?;
+
+    let cqe: cqueue::Entry = ring.completion().next().expect("cqueue is empty").into();
+    assert_eq!(cqe.user_data(), 0x26);
+
+    // send for recvmsg
+    send_stream.write_all(&[0x56u8; 1024])?;
+
+    // recvmsg
+    let mut msg = MaybeUninit::<libc::msghdr>::zeroed();
+    let sockaddr = socket2::SockAddr::from(recv_stream.local_addr()?);
+    let mut iovec = MaybeUninit::<libc::iovec>::zeroed();
+    unsafe {
+        let i = iovec.as_mut_ptr();
+        (*i).iov_len = 1024;
+        let p = msg.as_mut_ptr();
+        (*p).msg_name = sockaddr.as_ptr() as *const _ as *mut _;
+        (*p).msg_namelen = sockaddr.len();
+        (*p).msg_iov = iovec.as_mut_ptr();
+        (*p).msg_iovlen = 1;
+    }
+
+    let recvmsg_e = opcode::RecvMsg::new(recv_fd, msg.as_mut_ptr())
+        .buf_group(0xdeaf)
+        .build()
+        .flags(squeue::Flags::BUFFER_SELECT)
+        .user_data(0x27);
+
+    unsafe {
+        ring.submission()
+            .push(&recvmsg_e.into())
+            .expect("queue is full");
+    }
+
+    ring.submit_and_wait(1)?;
+
+    let cqe: cqueue::Entry = ring.completion().next().expect("cqueue is empty").into();
+    assert_eq!(cqe.user_data(), 0x27);
+    assert_eq!(cqe.result(), 1024);
+
+    let bid = cqueue::buffer_select(cqe.flags()).expect("no buffer id");
+    assert_eq!(bid, 0);
+    assert_eq!(&(recvmsg_buf[..]), &([0x56u8; 1024][..]));
+
+    // provide buf for readv
+    let mut readv_buf = [0u8; 512];
+    let provide_bufs_e = opcode::ProvideBuffers::new(readv_buf.as_mut_ptr(), 1024, 1, 0xdeb0, 0);
+
+    unsafe {
+        ring.submission()
+            .push(&provide_bufs_e.build().user_data(0x28).into())
+            .expect("queue is full");
+    }
+
+    ring.submit_and_wait(1)?;
+
+    let cqe: cqueue::Entry = ring.completion().next().expect("cqueue is empty").into();
+    assert_eq!(cqe.user_data(), 0x28);
+
+    // send for readv
+    send_stream.write_all(&[0x7bu8; 512])?;
+
+    // readv
+    let readv_e = opcode::Readv::new(
+        recv_fd,
+        &libc::iovec {
```

Comment:
This looks dangerous, it would be nice to put iovec definition on a separate line to clarify lifetime.

Reply:
That very interesting! Why is it dangerous but the compiler allows it?

Reply:
I think because a long time ago the ops were all deemed unsafe but the decision was made the API wouldn't be very pretty if all were explicitly defined as unsafe.

Reply:
The ops don't even have a safety warning. I guess one could say the submission comes with its own requirement of unsafe, but the unit tests don't include any safety comment reminding the reader how they are maintaining the safety contract. So they aren't a good example of proper use of the API, so it is too easy to make a mistake like this one.

Reply:
I'll make these changes, but I don't think we are learning all the lessons we could be from these things. Developers will continue to pass references to the ops without knowing the compiler isn't helping them like they are accustomed to from a maintained crate from `tokio-rs`.